### PR TITLE
C#: Sources for the `Dapper` database library

### PR DESCRIPTION
**Pull request by egregius313**
Adds sources for the `Dapper` database access layer library

### CODE REVIEW

File: csharp/ql/lib/ext/Dapper.model.yml [hunk 1]
```diff
@@ -58,3 +58,58 @@ extensions:
...
+      - ["Dapper", "SqlMapper", False, "QuerySingleOrDefaultAsync", "(System.Data.IDbConnection,System.String,System.Object,System.Data.IDbTransaction,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "ReturnValue", "database", "manual"]
+      - ["Dapper", "SqlMapper", False, "QuerySingleOrDefaultAsync", "(System.Data.IDbConnection,System.Type,System.String,System.Object,System.Data.IDbTransaction,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "ReturnValue", "database", "manual"]
+      - ["Dapper", "SqlMapper", False, "QuerySingleOrDefaultAsync<T>", "(System.Data.IDbConnection,System.String,System.Object,System.Data.IDbTransaction,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "ReturnValue", "database", "manual"]
+  - addsTo:
+      pack: codeql/csharp-all
+      extensible: summaryModel
+    data:
+      - ["Dapper", "SqlMapper", False, "AsList<T>", "(System.Collections.Generic.IEnumerable<T>)", "", "Argument[0]", "ReturnValue", "taint", "manual"]
```
Comment (tamasvajk) [inline csharp/ql/lib/ext/Dapper.model.yml:115]: Shouldn't this be `"Argument[0].Element", "ReturnValue.Element", "value", "manual"`?

**Review by tamasvajk — COMMENTED** [review]
I haven't checked the signatures in `sourceModel`. Could you generate these or did you have to manually compile this list? Do we need to differentiate between methods returning a single item (`ReturnValue`) vs an `IEnumerable<>` (`ReturnValue.Element`)?

File: csharp/ql/lib/ext/Dapper.model.yml (hunk 1 above)
Comment (hvitved) [reply csharp/ql/lib/ext/Dapper.model.yml:115]: Or, even better `"Argument[0].WithElement", "ReturnValue", "value", "manual"`

File: csharp/ql/lib/ext/Dapper.model.yml [hunk 2]
```diff
@@ -58,3 +58,58 @@ extensions:
...
       - ["Dapper", "SqlMapper", False, "QuerySingleOrDefaultAsync<T>", "(System.Data.IDbConnection,System.String,System.Object,System.Data.IDbTransaction,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "Argument[1]", "sql-injection", "manual"]
+  - addsTo:
+      pack: codeql/csharp-all
+      extensible: sourceModel
+    data:
+      - ["Dapper", "SqlMapper", False, "Query", "(System.Data.IDbConnection,System.String,System.Object,System.Data.IDbTransaction,System.Boolean,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "ReturnValue", "database", "manual"]
+      - ["Dapper", "SqlMapper", False, "Query", "(System.Data.IDbConnection,System.Type,System.String,System.Object,System.Data.IDbTransaction,System.Boolean,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "ReturnValue", "database", "manual"]
+      - ["Dapper", "SqlMapper", False, "Query<T>", "(System.Data.IDbConnection,System.String,System.Object,System.Data.IDbTransaction,System.Boolean,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "ReturnValue", "database", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/Dapper.model.yml:67]: Out of curiosity. Where did you find the signatures for the API? From the generated stubs or elsewhere?

Are there a missing summary for `Query<T>` (I am looking at the generated stubs).

File: csharp/ql/lib/ext/Dapper.model.yml [hunk 3]
```diff
@@ -58,3 +58,58 @@ extensions:
...
+      - ["Dapper", "SqlMapper", False, "Query<TFirst,TSecond,TReturn>", "(System.Data.IDbConnection,System.String,System.Func<TFirst,TSecond,TReturn>,System.Object,System.Data.IDbTransaction,System.Boolean,System.String,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "ReturnValue", "database", "manual"]
+      - ["Dapper", "SqlMapper", False, "Query<TFirst,TSecond,TThird,TFourth,TFifth,TReturn>", "(System.Data.IDbConnection,System.String,System.Func<TFirst,TSecond,TThird,TFourth,TFifth,TReturn>,System.Object,System.Data.IDbTransaction,System.Boolean,System.String,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "ReturnValue", "database", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/Dapper.model.yml:69]: Simply due to the complexity of the signatures, I think a wildcard should be sufficient. It seems unlikely that there are methods with a Query like name where the return value is not a database source.

File: csharp/ql/lib/ext/Dapper.model.yml (hunk 2 above)
Comment (michaelnebel) [reply csharp/ql/lib/ext/Dapper.model.yml:67]: Ah, maybe you looked at the "corresponding" sinks.

File: csharp/ql/lib/ext/Dapper.model.yml [hunk 4]
```diff
@@ -58,3 +58,58 @@ extensions:
...
+      - ["Dapper", "SqlMapper", False, "Query<TReturn>", "(System.Data.IDbConnection,System.String,System.Type[],System.Func<System.Object[],TReturn>,System.Object,System.Data.IDbTransaction,System.Boolean,System.String,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "ReturnValue", "database", "manual"]
+      - ["Dapper", "SqlMapper", False, "QueryAsync", "(System.Data.IDbConnection,System.String,System.Object,System.Data.IDbTransaction,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "ReturnValue", "database", "manual"]
+      - ["Dapper", "SqlMapper", False, "QueryAsync", "(System.Data.IDbConnection,System.Type,System.String,System.Object,System.Data.IDbTransaction,System.Boolean,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "ReturnValue", "database", "manual"]
+      - ["Dapper", "SqlMapper", False, "QueryAsync<T>", "(System.Data.IDbConnection,System.String,System.Object,System.Data.IDbTransaction,System.Nullable<System.Int32>,System.Nullable<System.Data.CommandType>)", "", "ReturnValue", "database", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/Dapper.model.yml:77]: It looks like all the generics have overloads. If you were inspired by the corresponding sinks models, I think they only have models for the methods where a string is provided as the SQL command.

**Comment by michaelnebel** [comment]
An overarching question:
Should we go through all modelled SQL sinks (and libraries) to ensure that we have corresponding sources for these?

File: csharp/ql/lib/ext/Dapper.model.yml (hunk 1 above)
Comment (michaelnebel) [reply csharp/ql/lib/ext/Dapper.model.yml:115]: I keep forgetting the intricate differences between these...

File: csharp/ql/lib/ext/Dapper.model.yml (hunk 2 above)
Comment (egregius313) [reply csharp/ql/lib/ext/Dapper.model.yml:67]: Yes, the source models I added were based on the sinks

File: csharp/ql/lib/ext/Dapper.model.yml (hunk 4 above)
Comment (egregius313) [reply csharp/ql/lib/ext/Dapper.model.yml]: Noted. I've moved the models over to using the wildcards.

File: csharp/ql/lib/ext/Dapper.model.yml (hunk 1 above)
Comment (egregius313) [reply csharp/ql/lib/ext/Dapper.model.yml]: @hvitved I'm not sure of the specifics of `.WithElement`, but the following test fails with the `.WithElement` model and succeeds with a `Argument[0].Element -> ReturnValue.Element` model? Is there a detail/subtlety I'm missing?

```csharp
public static void Bad04(string connectionString, string query)
{
    using (var connection = new SqlConnection(connectionString))
    {
        var results = connection.Query<object>(query).AsList();
        Sink(results[0]); // $ hasTaintFlow=line:43
    }
}
```

**Comment by egregius313** [comment]
> An overarching question: Should we go through all modelled SQL sinks (and libraries) to ensure that we have corresponding sources for these (not on this PR)?

I think it would be worthwhile. From my time looking over what we currently have modeled, several of the sinks are already covered (e.g. by our modeling of `DbDataReader`) or would be easy to model, similar to these Dapper models.

There are only a few packages we've modeled with `sql-injection` sinks and they mostly have fairly similar APIs.

**Comment by michaelnebel** [comment]
> > An overarching question: Should we go through all modelled SQL sinks (and libraries) to ensure that we have corresponding sources for these (not on this PR)?
> 
> I think it would be worthwhile. From my time looking over what we currently have modeled, several of the sinks are already covered (e.g. by our modeling of `DbDataReader`) or would be easy to model, similar to these Dapper models.
> 
> There are only a few packages we've modeled with `sql-injection` sinks and they mostly have fairly similar APIs.

That sounds sensible!
Do you intend to incorporate that in the description of the possible next steps?

**Review by michaelnebel — DISMISSED** [review]
LGTM!

**Comment by egregius313** [comment]
> That sounds sensible!
Do you intend to incorporate that in the issue describing the next steps?

Yes I can do that, I might put a note about it in our database modeling issue as well

**Comment by michaelnebel** [comment]
This PR needs to be re-based and the flow summaries test needs to be updated (due to https://github.com/github/codeql/pull/15940)

File: csharp/ql/lib/ext/Dapper.model.yml (hunk 1 above)
Comment (michaelnebel) [reply csharp/ql/lib/ext/Dapper.model.yml]: That doesn't work for some reason. I will open an issue for this.

**Comment by michaelnebel** [comment]
- DCA looks good.
- Updated the expected output of the source/sink test.
- Updated the AsList model to a `value` model.
- Opened an issue for the `.WithElement` problem.

**Review by michaelnebel — APPROVED** [review]
LGTM

**Pull request merged**